### PR TITLE
Run `ty` in lint CI

### PR DESCRIPTION
**Pull request by janosh**
with number of PRs, contributors and users increasing, it's more important to guard against easy-to-catch type errors. lessens burden on reviewers and allows to merge green PRs with more confidence

### CODE REVIEW

**Comment by thomasloux** [comment]
Didn't know this tool, nice to have. But there is a large amount of work to make ty happy with the current code

**Comment by orionarcher** [comment]
I support adding `ty` but it might be best to wait until the most recent of `MACE` is on PyPI, last version was released in August. With the most recent version of MACE, it was possible to get the number of type errors down to single digits. @janosh idk if you can bother Ilyas about it.

**Comment by orionarcher** [comment]
MACE finally [released v0.3.15](https://github.com/ACEsuit/mace/releases/tag/v0.3.15) so I'd say this is ready to move forward on

**Comment by janosh** [comment]
@orionarcher set a bunch of agents to work on this and coarsely reviewed their changes. looks good for a start but can definitely be polished. short on time for that. if anyone wants to have a go at that feel free to push to this branch! this will get blocked by merge conflicts quickly

**Comment by orionarcher** [comment]
This looks broadly good!

I am (CC is) gonna take a swing at this from scratch and see if I can get it to touch fewer LOC. It looks like its changing a decent amount of logic in addition to type signatures and I am curious if that is avoidable.

**Comment by janosh** [comment]
bit more context: started out with 670 type errors. agents bringing that down to 0 even with some hacks like ignoring a few important rules and files is a win. we can always reduce [bad config](https://github.com/TorchSim/torch-sim/pull/339/changes#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R198-R241) in subsequent PRs as time permits. but would good to have type checking on the rest of code base

if claude can bring the type error count to 0 with fewer changes, i'm all for it!

**Comment by janosh** [comment]
> I am (CC is) gonna take a swing at this from scratch and see if I can get it to touch fewer LOC.

@orionarcher how did it go?

meantime, i applied some manual polish to the agent edits, mostly aiming to cut any bloat and AI slop (see [e8a4cfa](https://github.com/TorchSim/torch-sim/pull/339/commits/e8a4cfab91d08ae1e47240add82652e63dff9f2b), [e16893a](https://github.com/TorchSim/torch-sim/pull/339/commits/e16893af58a301460dbf4fb1b927009cc8ee883d), [ff3bc7b](https://github.com/TorchSim/torch-sim/pull/339/commits/ff3bc7b96b068a82de108f07ea5245cdbe5295b7)). that worked in the sense that the diff line count in actual code files went down quite a bit. sadly i kept coming across what i think are bugs. i fixed them and added new tests to prevent future regressions. the tests in particular increased the net line count, despite my attempts to cut bloat but useful enough to warrant the lines imo:

- mixed-pbc bug in autobatching density scaling: `n_atoms_x_density` now uses the vectorized volume path only when **all** pbc axes are periodic (`pbc_all`), instead of allowing mixed-pbc cases to slip through and get incorrect scaling.  
  unless i'm missing something this one actually looks pretty serious! if anybody thinks it's not a bug, let me know. otherwise, my best guess for why this wasn't caught earlier is that nobody has used torch-sim for molecules or open surfaces yet ([commit](https://github.com/TorchSim/torch-sim/commit/ff3bc7b96b068a82de108f07ea5245cdbe5295b7), [diff](https://github.com/TorchSim/torch-sim/blob/ff3bc7b96b068a82de108f07ea5245cdbe5295b7/torch_sim/autobatching.py#L368-L404))

- `count_degrees_of_freedom()` now sums per-system DoF reduction before subtracting from scalar total. it looks like the function was assuming it's just dealing with a single system. this helper had zero coverage. constraints are tested broadly, but this function wasn’t directly exercised. i added 3 unit tests in `tests/test_constraints.py` for: no constraints, `FixAtoms` only, `FixCom` only, and mixed constraints across multi-system states, with exact DoF assertions. ([diff](https://github.com/TorchSim/torch-sim/blob/ff3bc7b96b068a82de108f07ea5245cdbe5295b7/torch_sim/constraints.py#L615-L625), [tests](https://github.com/TorchSim/torch-sim/blob/ff3bc7b96b068a82de108f07ea5245cdbe5295b7/tests/test_constraints.py#L661-L700))

- fire init/step now normalize scalar/tensor inputs consistently across systems ([commit](https://github.com/TorchSim/torch-sim/commit/ff3bc7b96b068a82de108f07ea5245cdbe5295b7), [diff](https://github.com/TorchSim/torch-sim/blob/ff3bc7b96b068a82de108f07ea5245cdbe5295b7/torch_sim/optimizers/fire.py#L26-L176))

- optional-neighbor-backend fallback robustness (alchemiops/vesin import and call guards): i think in some environments you’d get a clean ImportError and fallback. in others, you could hit a later “object not callable” style failure path, depending on exactly which optional packages are present.

File: pyproject.toml [hunk 1]
```diff
@@ -28,23 +28,27 @@ classifiers = [
...
     "psutil>=7.0.0",
+    "pymatgen>=2025.6.14",
     "pytest-cov>=6",
     "pytest>=8",
+    "spglib>=2.6",
 ]
 io = ["ase>=3.26", "phonopy>=2.37.0", "pymatgen>=2025.6.14"]
-symmetry = ["moyopy>=0.3", "spglib>=2.6"]
+symmetry = ["moyopy>=0.3"]
 mace = ["mace-torch>=0.3.14"]
```
Comment (orionarcher) [inline pyproject.toml]: We should probably pin mace to 0.3.15, I suspect ty will fail with earlier versions. It occurs to me that while we are running the tests with lowest and highest deps, I don't think we are doing the same for linting.

**Review by orionarcher — APPROVED** [review]
I didn't do a thorough line by line review but everything I saw LGTM. Just one comment. 

Also, one question, what versions of upstream software will `ty` look at?

**Comment by janosh** [comment]
> Also, one question, what versions of upstream software will `ty` look at?

`ty` uses whatever package versions are installed in the environment it runs in. we set no version pins in [.pre-commit-config.yaml](https://github.com/TorchSim/torch-sim/blob/d81bbf1fd46230403d9193ef0b3fc982f208739a/.pre-commit-config.yaml#L46), meaning we tell `prek` to install the latest versions of `ty`, `torch`, `ase`, `mace-torch`

https://github.com/TorchSim/torch-sim/blob/d81bbf1fd46230403d9193ef0b3fc982f208739a/.pre-commit-config.yaml#L46

File: pyproject.toml (hunk 1 above)
Comment (janosh) [reply pyproject.toml]: good catch, done in https://github.com/TorchSim/torch-sim/pull/339/changes/b79b8aacc3e46976d1ca870ecf5e4a5113d45210

File: torch_sim/models/particle_life.py [hunk 2]
```diff
@@ -124,7 +140,7 @@ def __init__(
             cutoff or 2.5 * sigma, dtype=self.dtype, device=self.device
         )
         self.epsilon = torch.tensor(epsilon, dtype=self.dtype, device=self.device)
-        self.beta = torch.tensor(0.3, dtype=self.dtype, device=self.device)
+        self.beta = torch.tensor(beta, dtype=self.dtype, device=self.device)
```
Comment (janosh) [inline torch_sim/models/particle_life.py:143]: was `beta` hard-coded for a reason? maybe @abhijeetgangan?

**Comment by janosh** [comment]
@CompRhys just noticed agents had removed the `due.dcite` decorators you'd added. `dcite` import was removed to satisfy the type checker. I think i restored them all:

**NVT integrators** (`torch_sim/integrators/nvt.py`):
- `nvt_langevin_step` — [`10.1098/rspa.2016.0138`](https://doi.org/10.1098/rspa.2016.0138) (Leimkuhler & Matthews, BAOAB Langevin)
- `nvt_nose_hoover_step` — [`10.1080/00268979600100761`](https://doi.org/10.1080/00268979600100761) (Martyna et al., Nose-Hoover chains)
- `nvt_vrescale_step` — [`10.1063/1.2408420`](https://doi.org/10.1063/1.2408420) (Bussi et al., velocity rescaling)

**NPT integrators** (`torch_sim/integrators/npt.py`):
- `npt_langevin_step` — [`10.1063/1.4901303`](https://doi.org/10.1063/1.4901303) (NPT Langevin)
- `npt_nose_hoover_step` — [`10.1080/00268979600100761`](https://doi.org/10.1080/00268979600100761) (Martyna et al., Nose-Hoover chains)
- `npt_crescale_anisotropic_step` — [`10.1063/5.0020514`](https://doi.org/10.1063/5.0020514) + [`10.3390/app12031139`](https://doi.org/10.3390/app12031139) (C-Rescale barostat)
- `npt_crescale_independent_lengths_step` — same two DOIs
- `npt_crescale_average_anisotropic_step` — same two DOIs
- `npt_crescale_isotropic_step` — [`10.1063/5.0020514`](https://doi.org/10.1063/5.0020514)

thanks for adding those! 👍

File: torch_sim/integrators/nvt.py [hunk 3]
```diff
@@ -179,9 +190,17 @@ def nvt_langevin_step(
...
+        gamma = 1 / (100 * dt)
+
+    if isinstance(gamma, float):
+        gamma = torch.tensor(gamma, device=device, dtype=dtype)
+
+    if isinstance(dt, float):
+        dt = torch.tensor(dt, device=device, dtype=dtype)
+
+    if isinstance(kT, float):
+        kT = torch.tensor(kT, device=device, dtype=dtype)
```
Comment (CompRhys) [inline torch_sim/integrators/nvt.py:203]: https://github.com/TorchSim/torch-sim/pull/466 :(

File: torch_sim/integrators/nvt.py [hunk 4]
```diff
@@ -577,15 +602,14 @@ def nvt_vrescale_init(
...
-    To seed the RNG set ``state.rng = seed`` before calling.
-
     Args:
         model: Neural network model that computes energies and forces.
             Must return a dict with 'energy' and 'forces' keys.
         state: Either a SimState object or a dictionary containing positions,
             masses, cell, pbc, and other required state vars
         kT: Temperature in energy units for initializing momenta,
             either scalar or with shape [n_systems]
+        seed: Random seed for reproducibility
```
Comment (CompRhys) [inline torch_sim/integrators/nvt.py:612]: seed kwarg erroneously reintroduced after being intentionally removed in https://github.com/TorchSim/torch-sim/pull/460

File: torch_sim/integrators/nvt.py [hunk 5]
```diff
@@ -596,16 +620,22 @@ def nvt_vrescale_init(
...
 
+    system_idx = state.system_idx
+    if system_idx is None:
+        raise ValueError("system_idx cannot be None for NVT integration")
     momenta = getattr(
         state,
         "momenta",
-        initialize_momenta(
-            state.positions, state.masses, state.system_idx, kT, state.rng
+        calculate_momenta(
```
Comment (CompRhys) [inline torch_sim/integrators/nvt.py:633]: the function name being reverted suggests all of #460 was reverted?

**Comment by CompRhys** [comment]
I think large parts of #460 and #466 were reverted here. #466 was just to try have cleaner code in some parts but #460 were real intentional removals of seed

File: torch_sim/integrators/nvt.py (hunk 5 above)
Comment (janosh) [reply torch_sim/integrators/nvt.py:633]: oops, thanks for spotting! i'll restore
